torchtitan/metrics.py

```diff
@@ -134,27 +180,47 @@ def build_metric_logger(
     job_config: JobConfig, parallel_dims: ParallelDims, tag: Optional[str] = None
 ):
     """
-    parallel_dims is used to determine the rank to log metrics from if 'tb_config.rank_0_only=True'.
-    In that case, `_get_metrics_rank` will be used to calculate which rank acts as 'rank 0'. This is
-    intended to allow logging from the 0th rank within the last pipeline stage group, in case pipeline
-    parallelism is enabled, without forcing logging from all ranks to capture loss information.
+    Builds a metric logger that can log to both TensorBoard and W&B (if enabled).
+    W&B support is optional and controlled via the metrics config.
+
+    Args:
+        job_config: Configuration object containing metrics settings
+        parallel_dims: Parallel dimensions configuration
+        tag: Optional tag to prefix all metrics
+
+    Returns:
+        MetricLogger instance configured based on the provided settings
     """
     dump_dir = job_config.job.dump_folder
     tb_config = job_config.metrics
     save_tb_folder = tb_config.save_tb_folder
-    # since we don't have run id, use current minute as the identifier
     datetime_str = datetime.now().strftime("%Y%m%d-%H%M")
     log_dir = os.path.join(dump_dir, save_tb_folder, datetime_str)
 
     enable_tb = tb_config.enable_tensorboard
+    enable_wandb = getattr(tb_config, "enable_wandb", False)
+    wandb_config = getattr(tb_config, "wandb_config", None)
+
     if enable_tb:
         logger.info(
             f"Metrics logging active. Tensorboard logs will be saved at {log_dir}"
         )
         if tb_config.rank_0_only:
-            enable_tb = torch.distributed.get_rank() == _get_metrics_rank(parallel_dims)
+            metrics_rank = _get_metrics_rank(parallel_dims)
+            enable_tb = torch.distributed.get_rank() == metrics_rank
+            enable_wandb = enable_wandb and (
+                torch.distributed.get_rank() == metrics_rank
+            )
         else:
             rank_str = f"rank_{torch.distributed.get_rank()}"
             log_dir = os.path.join(log_dir, rank_str)
 
-    return MetricLogger(log_dir, tag, enable_tb)
+    if enable_wandb and not WANDB_AVAILABLE:
+        logger.warning(
+            "W&B logging requested but wandb package is not installed. Continuing without W&B logging."
+        )
+        enable_wandb = False
+    elif enable_wandb:
```

Comment:
Just a nit: W&B logs it's own init statement, so having one here might be overkill.

Reply:
yeah good point, w&b is actually quite loud with emojis